osu.Game/Online/Leaderboards/Leaderboard.cs

```diff
@@ -127,137 +88,31 @@ protected override void LoadComplete()
 
-            if (api != null)
-            {
-                apiState.BindTo(api.State);
-                apiState.BindValueChanged(state =>
-                {
-                    switch (state.NewValue)
-                    {
-                        case APIState.Online:
-                        case APIState.Offline:
-                            if (IsOnlineScope)
-                                RefetchScores();
-
-                            break;
-                    }
-                });
-            }
-
-            RefetchScores();
+            LeaderboardScoresProvider.State.BindValueChanged(state => onStateChange(state.NewValue));
         }
 
-        /// <summary>
-        /// Perform a full refetch of scores using current criteria.
-        /// </summary>
-        public void RefetchScores() => Scheduler.AddOnce(refetchScores);
-
-        /// <summary>
-        /// Clear all scores from the display.
-        /// </summary>
-        public void ClearScores()
+        private void onStateChange(LeaderboardState state)
         {
-            cancelPendingWork();
-            SetScores(null);
-        }
-
-        /// <summary>
-        /// Call when a retrieval or display failure happened to show a relevant message to the user.
-        /// </summary>
-        /// <param name="state">The state to display.</param>
-        protected void SetErrorState(LeaderboardState state)
-        {
-            switch (state)
-            {
-                case LeaderboardState.NoScores:
-                case LeaderboardState.Retrieving:
-                case LeaderboardState.Success:
-                    throw new InvalidOperationException($"State {state} cannot be set by a leaderboard implementation.");
-            }
-
-            Debug.Assert(!scores.Any());
-
-            setState(state);
-        }
-
-        /// <summary>
-        /// Call when retrieved scores are ready to be displayed.
-        /// </summary>
-        /// <param name="scores">The scores to display.</param>
-        /// <param name="userScore">The user top score, if any.</param>
-        protected void SetScores(IEnumerable<TScoreInfo>? scores, TScoreInfo? userScore = default)
-        {
-            this.scores.Clear();
-            if (scores != null)
-                this.scores.AddRange(scores);
-
-            // Non-delayed schedule may potentially run inline (due to IsMainThread check passing) after leaderboard  is disposed.
-            // This is guarded against in BeatmapLeaderboard via web request cancellation, but let's be extra safe.
-            if (!IsDisposed)
-            {
-                // Schedule needs to be non-delayed here for the weird logic in refetchScores to work.
-                // If it is removed, the placeholder will be incorrectly updated to "no scores" rather than "retrieving".
-                // This whole flow should be refactored in the future.
-                Scheduler.Add(applyNewScores, false);
```

Comment:
I think this could now be removed, as while testing I didn't encounter such incorrect behaviour.